lib/spack/spack/build_systems/meson.py

```diff
@@ -0,0 +1,185 @@
+# published by the Free Software Foundation) version 2.1, February 1999.
+#
+# This program is distributed in the hope that it will be useful, but
+# WITHOUT ANY WARRANTY; without even the IMPLIED WARRANTY OF
+# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the terms and
+# conditions of the GNU Lesser General Public License for more details.
+#
+# You should have received a copy of the GNU Lesser General Public
+# License along with this program; if not, write to the Free Software
+# Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA 02111-1307 USA
+##############################################################################
+
+import inspect
+import os
+
+from llnl.util.filesystem import working_dir
+from spack.directives import depends_on, variant
+from spack.package import PackageBase, run_after
+
+
+class MesonPackage(PackageBase):
+    """Specialized class for packages built using Meson
+
+    For more information on the Meson build system, see:
+    https://mesonbuild.com/
+
+    This class provides three phases that can be overridden:
+
+        1. :py:meth:`~.MesonPackage.meson`
+        2. :py:meth:`~.MesonPackage.build`
+        3. :py:meth:`~.MesonPackage.install`
+
+    They all have sensible defaults and for many packages the only thing
+    necessary will be to override :py:meth:`~.MesonPackage.meson_args`.
+    For a finer tuning you may also override:
+
+        +-----------------------------------------------+--------------------+
+        | **Method**                                    | **Purpose**        |
+        +===============================================+====================+
+        | :py:meth:`~.MesonPackage.root_mesonlists_dir` | Location of the    |
+        |                                               | root MesonLists.txt|
+        +-----------------------------------------------+--------------------+
+        | :py:meth:`~.MesonPackage.build_directory`     | Directory where to |
+        |                                               | build the package  |
+        +-----------------------------------------------+--------------------+
+
+
+    """
+    #: Phases of a Meson package
+    phases = ['meson', 'build', 'install']
+    #: This attribute is used in UI queries that need to know the build
+    #: system base class
+    build_system_class = 'MesonPackage'
+
+    build_targets = []
+    install_targets = ['install']
+
+    build_time_test_callbacks = ['check']
+
+    variant('buildtype', default='release',
+            description='Meson build type',
+            values=('plain', 'debug', 'debugoptimized', 'release', 'minsize'))
+
+    depends_on('meson', type='build')
+    depends_on('ninja', type='build')
+
+    @property
+    def archive_files(self):
+        """Files to archive for packages based on Meson"""
+        return [os.path.join(self.build_directory, 'meson-logs/meson-log.txt')]
+
+    @property
+    def root_mesonlists_dir(self):
```

Comment:
Does Meson do the thing that CMake does where you have recursive CMakeLists.txt files? If that's the case, then this is fine, but I would rename it to `root_meson_build_dir`. If not, I would rename this `meson_build_dir` or `mesonbuild_directory` to be more in line with `AutotoolsPackage`. The recursion is the only reason it was called `root`, to distinguish that there can be many directories containing the file.

Reply:
It allows subprojects.  I have not seen it yet, but this build system is new.